contracts/contracts/interfaces/staking-manager.md

```diff
@@ -0,0 +1,109 @@
+
+The contract should verify these:
+
+- The node is already registered as a validator
+- The address that registered the node is the same with the address that is removing the validator
+- (Optional) Validator being removed after minimum staking period
+
+After verification contract should perform the following:
+
+- Send a warp message to P-Chain for `SetSubnetValidatorWeightTx{messageID: relatedRegisterValidatorMessageID, nonce: MaxUint64, Weight: 0, }` transaction
+- Stop accruing rewards for the validator
+
+Once the `SetSubnetValidatorWeightTx` is accepted, a warp message `InvalidValidatorRegisterMessage` will be relayed to the contract with `messageID` of the request. The message represents a messageID will not possibly become a valid validator. This essentialy means the message was not valid in the first place, or the validator has removed from the set. The InvalidValidatorRegisterMessage can either be a result of `removeValidator` from the contract or it can be originated directly from P-chain via `ExitValidatorSetTx`. In either case this ultimately tells the contract that a validator was evicted. See related ACP-77 section [here](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#exitvalidatorsettx).
+
+The contract should verify these in `receiveRemoveValidator`:
+
+- Verified warp message
+- The typecheck on the warp message (== InvalidValidatorRegisterMessage)
+- messageID is not empty
+
+At this point the contract should unlock all funds that were locked for the validator and issue locked rewards. If the validator was not removed by the contract itself, the contract should either give partial rewards or no rewards at all. This is because we don't know when the validator was removed.
+
+Rewards should not be unlocked or sent to the staker at this point. This is because we need to first perform a uptime check to determine the rewards.
+
+TODO: should we give a partial reward in case the validator got removed from P-chain (balance drained)? Possibly by timestamping the `InvalidValidatorRegisterMessage`.
+
+TODO: should we add a cooldown period for validator removal/readd?
+
+## Uptime Based Rewards
+
+The manager contract should reward based on uptimes of validators. VMs should be able to track precise uptimes of validators. The manager contract should be able to query the uptime of a validator and reward based on the uptime.
+
+TODO: Design currently is not finalized. We need to discuss how to track uptimes and how to report them to the manager contract. Some ideas:
+
+1- If a validator becomes ineligible permanetly due to low uptime, a warp message should be sent to the manager contract.
+2- Once a validator is removed, a warp message that indicates the uptime is enough will be sent to the contract.
+3- A message will be aggregated upon a request from the manager contract (e.g via an event), then the message will be sent to the manager contract.
```

Comment:
I feel like if the manager contract can signal for aggregation of a message which target a uptime specified by the contract; it can implement many different use cases. I.e contract can ask about >%80 uptime message to be signed and if so it can use it to rewarding. In another case it might ask for <%50 uptime and if signed can slash. A simply dynamic request for >%x itself is very helpful for contracts to determine their min uptime.

Reply:
Pseudocode :
```
  constant  threshold = 50

function receiveRegisterMessageInvalid(uint32 messageIndex) external {
  ...
  if (validator.redeemedAt != 0) {
    uint256 reward = calculateReward(
      validator.weight,
      validator.startedAt,
      validator.redeemedAt
    );
    uptimeID = calculateID(validationID, threshold, reward)
    lockedRewards[uptimeID] = {validator.rewardAddress, reward}
    emit UptimeRequest(uptimeID, threshold, validationID, validator.nodeID)
   ...
}

function receiveUptime(uint32 messageIndex) external{
  (WarpMessage memory warpMessage, bool success) = WARP_MESSENGER.getVerifiedWarpMessage(messageIndex);
  require(success, "ExampleValidatorManager: invalid warp message");
  uptimeMessage = decode(warpMessage)
  uptimeID = uptimeMessage.ID
  reward = lockedRewards[uptimeID] 
  delete lockedRewards[uptimeID] 
  stakingToken.transfer(reward.address, reward.amount)
  ...
}
```

Reply:
I prefer this flexible alternative, do we plan to support only >= and < operators?

Reply:
In most of the cases I don't think we even need to support "<". Most of the time it will be enough to check if a validator has enough uptime. However I also don't think it would be too complicated to support "<". I think these both can cover almost all use cases.